### PR TITLE
Allow more generic SSL verification (fixes #244)

### LINKED ISSUE #244: Provide a hook for arbitrary SSL verification methods
Currently we use a `VerifyCallback`, however there's a lot of in built verification in `SslContext` (eg you can just provide it with a root CA file and it does the rest).  Making `HttpConnector` take a `fn(&mut SslContext)` would be more useful (one can still attach verify callbacks directly, so no functionality is lost)

I have a pre-unboxed closure patch [here](https://github.com/Manishearth/hyper/commit/a815cddecfa1d1ac86b436bf1456c87b6ed06552), if its design feels okay I can move it to unboxed closures.

### PR DESCRIPTION
**Pull request by Manishearth**

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 8e44720395755f57d44049e2ea2bf6806bf679d3
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 69d3aeba1f1793aed3a1b42b3180f6fe293e9135
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/hyperium/hyper/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by Manishearth** [comment]
This doesn't use unboxed closures yet, because I don't like having to specify a dummy unboxed closure every time. I tried using default type params to get around it, but it didn't work too well and I got [an ICE](https://github.com/rust-lang/rust/issues/21139). Working on a different design now.

**Comment by seanmonstar** [comment]
I don't understand what you mean. I'd likely let set_ssl_verifier be generic over Fn, and then store it as  Box<Fn>.

File: src/net.rs [hunk 1]
```diff
@@ -325,7 +328,9 @@ impl NetworkConnector for HttpConnector {
                 debug!("https scheme");
                 let stream = try!(TcpStream::connect(addr));
                 let mut context = try!(SslContext::new(Sslv23).map_err(lift_ssl_error));
-                self.0.as_ref().map(|cb| context.set_verify(SslVerifyPeer, Some(*cb)));
+                if let Some(ref v) = self.0 {
```
Comment (reem) [inline src/net.rs]: Can we use a more descriptive name like `verifier`? It's a small change but it brings clarity.

**Comment by Manishearth** [comment]
@seanmonstar Oh, I thought "boxed unboxed closures" didn't work yet

File: src/net.rs (hunk 1 above)
Comment (Manishearth) [inline src/net.rs]: (fixed)

**Pull request merged**